extensions/sockets/src/main/java/org/mule/module/socket/api/connection/udp/UdpListenerConnection.java

```diff
@@ -0,0 +1,98 @@
+import org.mule.runtime.extension.api.runtime.MessageHandler;
+import org.mule.runtime.extension.api.runtime.source.Source;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.SocketTimeoutException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class UdpListenerConnection extends AbstractUdpConnection implements ListenerConnection
+{
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(UdpListenerConnection.class);
+    private boolean wasDisconnected = false;
+
+    public UdpListenerConnection(UdpSocketProperties socketProperties, String host, Integer port) throws ConnectionException
+    {
+        super(socketProperties, host, port);
+    }
+
+    @Override
+    public void connect() throws ConnectionException
+    {
+        socket = newSocket(host, port);
+        configureConnection(socket, socketProperties);
+    }
+
+    @Override
+    public void disconnect()
+    {
+        wasDisconnected = true;
+        super.disconnect();
+    }
+
+    /**
+     * One worker is created per received package. If the other end of the connection is awaiting
+     * for a response, one will be sent but not from the same listener socket the source has. The response
+     * will be sent from a new different {@link DatagramSocket} bound to a port choose by the system.
+     * <p>
+     * The new {@link DatagramSocket} used for responding has the same configuration settings than the
+     * listener socket used in the {@link Source}.
+     *
+     * @throws ReadingTimeoutException if the socket reached timeout while awaiting the arrival of a new package.
+     * @throws ConnectionException     if the socket was close while blocked on {@link DatagramSocket#receive(DatagramPacket)} method
+     *                                 and the close was not intentionally done by the {@link UdpListenerConnection#disconnect()} method.
+     * @throws IOException             if the socket was close while blocked on {@link DatagramSocket#receive(DatagramPacket)} method.
+     */
+    @Override
+    public SocketWorker listen(MuleContext muleContext, MessageHandler<InputStream, SocketAttributes> messageHandler) throws IOException, ConnectionException
+    {
+        DatagramPacket packet = createPacket(socketProperties.getReceiveBufferSize());
+
+        try
+        {
+            socket.receive(packet);
+        }
+        catch (SocketTimeoutException e)
+        {
+            throw new ReadingTimeoutException("UDP Source timed out while awaiting for new packages", e);
+        }
+        catch (IOException e)
+        {
+            if (wasDisconnected)
+            {
+                // socket was not intentionally closed by the connection provider
+                LOGGER.debug("UDP listener socket has been gracefully closed");
+                throw new ConnectionException("An error occurred while listening for new UDP packets", e);
```

Comment:
if it was gracefully stopped why do you request a reconnection by throwing this exception? Shouldn't you just log that you received this exception after being stopped?